File: plugins/geo_fencing/android/src/main/kotlin/com/example/geo_fencing/database/AppDatabase.kt
package com.example.geo_fencing.database

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.example.geo_fencing.PluginParams

class AppDatabase private constructor(context: Context, databasePath: String, version: Int) :
    SQLiteOpenHelper(context, databasePath, null, version, null) {


    companion object {


        @Volatile
        private var INSTANCE: AppDatabase? = null

        fun getInstance(): AppDatabase {
            if (INSTANCE == null) {
                throw Exception("database not initialized. call init method first")
            }
            return INSTANCE!!
        }

        fun init(context: Context, databasePath: String, version: Int): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                val dbPath = "${databasePath}/geo_fencing.db"
                INSTANCE ?: AppDatabase(context.applicationContext, dbPath, version)
                    .also { INSTANCE = it }
            }
        }
    }

    override fun onCreate(db: SQLiteDatabase?) {
        db?.execSQL(locationTableQuery)
        db?.execSQL(trackerTableQuery)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {}

    private val locationTableQuery
        get() = """
            CREATE TABLE IF NOT EXISTS ${PluginParams.LOCATION_TABLE} (
                ${LocationDao.COLUMN_ID} INTEGER PRIMARY KEY AUTOINCREMENT,
                ${LocationDao.COLUMN_NAME} TEXT NOT NULL,
                ${LocationDao.COLUMN_LONGITUDE} NUMERIC DEFAULT 0.0,
                $${LocationDao.COLUMN_LATITUDE} NUMERIC DEFAULT 0.0
            )
        """.trimIndent()

    private val trackerTableQuery
        get() = """
            CREATE TABLE IF NOT EXISTS ${PluginParams.TRACKER_TABLE} (
                $${TrackerDao.COLUMN_ID} INTEGER PRIMARY KEY AUTOINCREMENT,
                $${TrackerDao.COLUMN_LOCATION_ID} INTEGER NOT NULL,
                $${TrackerDao.COLUMN_FROM} timestamp DEFAULT CURRENT_TIMESTAMP,
                $${TrackerDao.COLUMN_TO} timestamp DEFAULT NULL
            )
        """.trimIndent()


}

File: plugins/geo_fencing/ios/Classes/database/AppDatabase.swift
//Made by Vahid Rajabi

import Foundation
import SQLite3

class AppDatabase {
    static let shared = AppDatabase()
    private var db: OpaquePointer?
    private var databaseInitialized = false
    
    private init() {}
    
    func initialize(path: String, version: Int) {
        let dbPath = "\(path)/geo_fencing.db"
        
        guard sqlite3_open(dbPath, &db) == SQLITE_OK else {
            print("Error opening database")
            return
        }
        
        createTables()
        databaseInitialized = true
    }
    
    var database: OpaquePointer? {
        if !databaseInitialized {
            fatalError("Database not initialized. Call initialize method first.")
        }
        return db
    }
    
    private func createTables() {
        guard let locationTable = PluginParams.LOCATION_TABLE,
              let trackerTable = PluginParams.TRACKER_TABLE else {
            return
        }
        
        let locationTableQuery = """
            CREATE TABLE IF NOT EXISTS \(locationTable) (
                \(LocationDao.COLUMN_ID) INTEGER PRIMARY KEY AUTOINCREMENT,
                \(LocationDao.COLUMN_NAME) TEXT NOT NULL,
                \(LocationDao.COLUMN_LONGITUDE) NUMERIC DEFAULT 0.0,
                \(LocationDao.COLUMN_LATITUDE) NUMERIC DEFAULT 0.0
            )
        """
        
        let trackerTableQuery = """
            CREATE TABLE IF NOT EXISTS \(trackerTable) (
                \(TrackerDao.COLUMN_ID) INTEGER PRIMARY KEY AUTOINCREMENT,
                \(TrackerDao.COLUMN_LOCATION_ID) INTEGER NOT NULL,
                \(TrackerDao.COLUMN_FROM) TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                \(TrackerDao.COLUMN_TO) TIMESTAMP DEFAULT NULL
            )
        """
        
        executeQuery(query: locationTableQuery)
        executeQuery(query: trackerTableQuery)
    }
    
    func executeQuery(query: String) {
        var statement: OpaquePointer?
        
        if sqlite3_prepare_v2(db, query, -1, &statement, nil) == SQLITE_OK {
            if sqlite3_step(statement) != SQLITE_DONE {
                print("Error executing query: \(query)")
            }
        } else {
            print("Error preparing query: \(query)")
        }
        
        sqlite3_finalize(statement)
    }
    
    deinit {
        if db != nil {
            sqlite3_close(db)
        }
    }
}
